// @stoplight/spectral-formatters/dist/index.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.codeClimate = exports.sarif = exports.githubActions = exports.pretty = exports.markdown = exports.teamcity = exports.stylish = exports.text = exports.junit = exports.json = exports.html = void 0;
var index_1 = require("./index");
Object.defineProperty(exports, "html", { enumerable: true, get: function () { return index_1.html; } });
Object.defineProperty(exports, "json", { enumerable: true, get: function () { return index_1.json; } });
Object.defineProperty(exports, "junit", { enumerable: true, get: function () { return index_1.junit; } });
Object.defineProperty(exports, "text", { enumerable: true, get: function () { return index_1.text; } });
Object.defineProperty(exports, "stylish", { enumerable: true, get: function () { return index_1.stylish; } });
Object.defineProperty(exports, "teamcity", { enumerable: true, get: function () { return index_1.teamcity; } });
Object.defineProperty(exports, "markdown", { enumerable: true, get: function () { return index_1.markdown; } });
var pretty_1 = require("./pretty");
Object.defineProperty(exports, "pretty", { enumerable: true, get: function () { return pretty_1.pretty; } });
var github_actions_1 = require("./github-actions");
Object.defineProperty(exports, "githubActions", { enumerable: true, get: function () { return github_actions_1.githubActions; } });
var sarif_1 = require("./sarif");
Object.defineProperty(exports, "sarif", { enumerable: true, get: function () { return sarif_1.sarif; } });
var code_climate_1 = require("./code-climate");
Object.defineProperty(exports, "codeClimate", { enumerable: true, get: function () { return code_climate_1.codeClimate; } });
